### PR TITLE
Add type layout stress mode to crossgen2/runtime

### PR DESCRIPTION
**Pull request by davidwrighton**
- New command line parameter --verify-type-and-field-layout
  - Intended for use diagnosing bugs in crossgen2 only, not for general
  production use
- New r2r precode fixups Verify_TypeLayout and Verify_FieldOffset
  - These allows the compiler to specify that a particular type/field
  layout must match with the runtime calculation.
  - If there is a mismatch, the runtime will failfast, even on release
  builds
- Add use of new switch to crossgen2smoke test
- Fix encoding bug in FieldFixupSignature. Check_FieldOffset and
Verify_FieldOffset use unsigned offsets, not signed offsets

### CODE REVIEW

**Review by nattress — APPROVED** [review]
LGTM - Thanks for adding this

**Comment by davidwrighton** [comment]
@dotnet/crossgen-contrib This is the first PR for this. Once this is in, I will start trying to run our larger test runs with the switch enabled and see how clean we are.

File: src/coreclr/src/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs [hunk 1]
```diff
@@ -1062,6 +1068,13 @@ private void getFieldInfo(ref CORINFO_RESOLVED_TOKEN pResolvedToken, CORINFO_MET
                     else
                     if (helperId != ReadyToRunHelperId.Invalid)
                     {
+                        if (_compilation.SymbolNodeFactory.VerifyTypeAndFieldLayout)
+                        if (_compilation.SymbolNodeFactory.VerifyTypeAndFieldLayout)
```
Comment (mangod9) [inline src/coreclr/src/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs:1072]: copy-paste issue?

Comment (davidwrighton) [reply src/coreclr/src/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs:1072]: Lol, yes. Conveniently, it doesn't have an impact on the correctness, but it is a bit ugly  :)

Comment (mangod9) [reply src/coreclr/src/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs:1072]: wonder if the compiler optimizes it out :)

**Review by mangod9 — APPROVED** [review]

**Pull request merged**